src/database/accessors/loadProposalBundle.ts

```diff
@@ -1,16 +1,222 @@
+        opportunityId: proposalPiece.proposal_opportunityId,
+        externalId: proposalPiece.proposal_externalId,
+        createdAt: proposalPiece.proposal_createdAt,
+      },
+    ],
+    [],
+  ).filter(filterByDistinctId)
+);
+
+const joinProposalFieldValuesWithApplicationFormFields = (
+  proposalFieldValues: ProposalFieldValue[],
+  applicationFormFields: ApplicationFormField[],
+): ProposalFieldValue[] => (
+  proposalFieldValues.map((proposalFieldValue) => ({
+    ...proposalFieldValue,
+    applicationFormField: applicationFormFields.find(
+      (applicationFormField) => (
+        applicationFormField.id === proposalFieldValue.applicationFormFieldId
+      ),
+    ),
+  }))
+);
+
+const joinProposalVersionsWithProposalFieldValues = (
+  proposalVersions: ProposalVersion[],
+  proposalFieldValues: ProposalFieldValue[],
+): ProposalVersion[] => (
+  proposalVersions.map((proposalVersion) => ({
+    ...proposalVersion,
+    fieldValues: proposalFieldValues.filter(
+      (proposalFieldValue) => proposalFieldValue.proposalVersionId === proposalVersion.id,
+    ),
+  }))
+);
+
+const joinProposalsWithProposalVersions = (
+  proposals: Proposal[],
+  proposalVersions: ProposalVersion[],
+): Proposal[] => (
+  proposals.map((proposal) => ({
+    ...proposal,
+    versions: proposalVersions.filter(
+      (proposalVersion) => proposalVersion.proposalId === proposal.id,
+    ),
+  }))
+);
+
+export const joinProposalFieldValuesToProposalVersion = (
+  proposalVersion: ProposalVersion,
+  values: ProposalFieldValue[],
+): ProposalVersion => {
+  const newVersion = structuredClone(proposalVersion);
+  values.forEach((proposalFieldValue) => {
+    if (newVersion.fieldValues === undefined) {
+      newVersion.fieldValues = [];
+    }
+    if (proposalFieldValue.proposalVersionId === newVersion.id) {
+      newVersion.fieldValues.push(proposalFieldValue);
+    }
+  });
+  return newVersion;
+};
+
 export const loadProposalBundle = async (
   queryParameters: TinyPgParams,
-): Promise<Bundle<Proposal>> => loadBundle(
-  'proposals.selectWithPagination',
-  queryParameters,
-  'proposals',
-  isProposal,
-);
+): Promise<Bundle<Proposal>> => {
+  // See https://github.com/typescript-eslint/typescript-eslint/issues/1824
+  /* eslint-disable @typescript-eslint/indent */
+  const unprocessedBundle = await loadBundle<ProposalPiece
+    | ProposalVersionPiece
+    | ProposalFieldValuePiece
+    | ApplicationFormFieldPiece>(
+  /* eslint-enable @typescript-eslint/indent */
+    'proposals.selectWithPagination',
+    queryParameters,
+    'proposals',
+  );
+  const deepProposalPieces = unprocessedBundle.entries;
+  const applicationFormFields = extractApplicationFormFields(
+    deepProposalPieces.filter(isApplicationFormFieldPiece),
+  );
+  const proposalFieldValues = joinProposalFieldValuesWithApplicationFormFields(
+    extractProposalFieldValues(
+      deepProposalPieces.filter(isProposalFieldValuePiece),
```

Comment:
I think I understand why the `filter`ing is done here - to get better type hints, right? - but it'd kinda be nice if that could happen in the `extract`or functions.

Reply:
It's required / the filter is a type guard.

The alternative would be to have the extract functions take in a `DeepProposalPiece` but then they would not be re-usable across other deep objects.

Reply:
Yeah, thank you for articulating that more clearly - that's what I was getting at. Kind of unfortunate.